src/lint/lint.ts

```diff
@@ -0,0 +1,400 @@
+      }
+      for (let rhs of rhses) {
+        if (!originalRhses.some(o => rhsMatches(rhs, o))) {
+          let { line, column } = getLocationInGrammar(rhs.pos);
+          lintingErrors.push({
+            line,
+            column,
+            message: `Could not find a production matching RHS in ${type}`,
+          });
+        }
+
+        if (rhs.kind === SyntaxKind.RightHandSide) {
+          (function noGrammarRestrictions(s: SymbolSpan | undefined) {
+            if (s === undefined) {
+              return;
+            }
+            if (s.symbol.kind === SyntaxKind.NoSymbolHereAssertion) {
+              let { line, column } = getLocationInGrammar(s.symbol.pos);
+              lintingErrors.push({
+                line,
+                column,
+                message: `Productions referenced in ${type}s should not include "no LineTerminator here" restrictions`,
+              });
+            }
+            // We could also enforce that lookahead restrictions are absent, but in some cases they actually do add clarity, so we just don't enforce it either way.
+
+            noGrammarRestrictions(s.next);
+          })(rhs.head);
+
+          if (rhs.constraints !== undefined) {
+            let { line, column } = getLocationInGrammar(rhs.constraints.pos);
+            lintingErrors.push({
+              line,
+              column,
+              message: `Productions referenced in ${type}s should not be gated on grammar parameters`,
+            });
+          }
+        }
+      }
+
+      // Filter out unused parameter errors for which the parameter is actually used in an SDO or Early Error
+      if (unusedParameterErrors.has(name)) {
+        let paramToError = unusedParameterErrors.get(name)!;
+        for (let [paramName, error] of paramToError) {
+          // This isn't the most elegant check, but it works.
+          if (rulesEles.some(r => r.innerHTML.indexOf('[' + paramName + ']') !== -1)) {
+            paramToError.delete(paramName);
+            // Yes, there's definitely big-O faster ways of doing this, but in practice this is probably faster for the sizes we will encounter.
+            lintingErrors = lintingErrors.filter(e => e !== error);
+          }
+        }
+      }
+    }
+  }
+
+  // *******************
+  // Enforce algorithm-specific linting rules
+
+  for (let algorithm of algorithms) {
+    let element = algorithm.element;
+    let location = getLocation(dom, element);
+
+    let reporter: Reporter = ({
+      line,
+      column,
+      message,
+    }: {
+      line: number;
+      column: number;
+      message: string;
+    }) => {
+      let trueLine = location.startTag.line + line - 1; // both jsdom and ecmarkdown have 1-based line numbers, so if we just add we are off by one
+      let trueCol = column;
+      if (line === 0) {
+        trueCol +=
+          location.startTag.col + (location.startTag.endOffset - location.startTag.startOffset);
+      } else {
+        trueCol += 1;
+      }
```

Comment:
This reads better to me.

```suggestion
      let trueCol = line === 0
        ? location.startTag.col + (location.startTag.endOffset - location.startTag.startOffset) + column
        : column + 1;
```

I'm also a little concerned about adding 1 only in the case that it's not the first line. Why is that? Also, I thought the comment above said the lines are 1-based, but we're checking for line `0`?

Reply:
> I'm also a little concerned about adding 1 only in the case that it's not the first line. Why is that?

Because jsdom's columns are 1-based, so if we include `location.startTag.col` we have already included the 1.

> Also, I thought the comment above said the lines are 1-based, but we're checking for line `0`?

Nice catch, fixed (and tested).

Also added more comprehensive comments to the places we do location adjustments. The four tools we're using have three different behaviors:

- grammarkdown is 0-based lines and 0-based columns
- ecmarkdown is 1-based lines and 0-based columns (which we did to match eslint)
- jsdom and this linter are 1-based for both (to match editors)